src/btree/bt_sync_obsolete.c

```diff
@@ -291,3 +290,324 @@ __wt_sync_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *parent)
+ *     Check and perform checkpoint cleanup on the uri.
+ */
+static int
+__checkpoint_cleanup_walk_btree(WT_SESSION_IMPL *session, const char *uri)
+{
+    WT_BTREE *btree;
+    WT_DECL_RET;
+    WT_REF *ref;
+    uint32_t flags;
+
+    ref = NULL;
+    flags = WT_READ_NO_EVICT;
+
+    /*
+     * To reduce the impact of checkpoint cleanup on the running database, it operates only on the
+     * dhandles that are already opened.
+     */
+    WT_WITHOUT_DHANDLE(session,
+      WT_WITH_HANDLE_LIST_READ_LOCK(session, (ret = __wt_conn_dhandle_find(session, uri, NULL))));
+    if (ret == WT_NOTFOUND)
+        return (0);
+
+    /* Open a handle for processing. */
+    ret = __wt_session_get_dhandle(session, uri, NULL, NULL, 0);
+    if (ret != 0)
+        WT_RET_MSG(session, ret, "%s: unable to open handle%s", uri,
+          ret == EBUSY ? ", error indicates handle is unavailable due to concurrent use" : "");
+
+    btree = S2BT(session);
+    /* There is nothing to do on an empty tree. */
+    if (btree->root.page == NULL)
+        goto err;
+
+    /*
+     * FLCS pages cannot be discarded and must be rewritten as implicitly filling in missing chunks
+     * of FLCS namespace is problematic.
+     */
+    if (btree->type == BTREE_COL_FIX)
+        goto err;
+
+    /* Walk the tree. */
+    while ((ret = __wt_tree_walk_custom_skip(
+              session, &ref, __checkpoint_cleanup_page_skip, NULL, flags)) == 0 &&
+      ref != NULL) {
+        if (F_ISSET(ref, WT_REF_FLAG_INTERNAL))
+            WT_WITH_PAGE_INDEX(session, ret = __checkpoint_cleanup_obsolete_cleanup(session, ref));
+        WT_ERR(ret);
+
+        /* Check if we're quitting. */
+        if (!__checkpoint_cleanup_run_chk(session))
+            break;
+    }
+
+err:
+    /* On error, clear any left-over tree walk. */
+    WT_TRET(__wt_page_release(session, ref, flags));
+    WT_TRET(__wt_session_release_dhandle(session));
+    return (ret);
+}
+
+/*
+ * __checkpoint_cleanup_int --
+ *     Internal function to perform checkpoint cleanup of all eligible files.
+ */
+static int
+__checkpoint_cleanup_int(WT_SESSION_IMPL *session)
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    const char *uri;
+
+    WT_RET(__wt_metadata_cursor(session, &cursor));
+    while ((ret = cursor->next(cursor)) == 0) {
+        WT_ERR(cursor->get_key(cursor, &uri));
+
+        /* Ignore non-btree objects as well as the metadata file. */
+        if (!WT_BTREE_PREFIX(uri) || strcmp(uri, WT_METAFILE_URI) == 0)
+            continue;
+
+        ret = __checkpoint_cleanup_walk_btree(session, uri);
+        if (ret == ENOENT || ret == EBUSY) {
+            __wt_verbose_debug1(session, WT_VERB_CHECKPOINT_CLEANUP,
+              "%s: skipped performing checkpoint cleanup because the file %s", uri,
+              ret == ENOENT ? "does not exist" : "is busy");
+            continue;
+        }
+        WT_ERR(ret);
+
+        /* Wait for 5 seconds before proceeding with another table. */
```

Comment:
Should we make that `5` seconds a constant somewhere?
Can this thread be event-driven and triggered whenever there is work to do?

Reply:
Furthermore, do we still want to have a cursor opened on the metadata file while sleeping? Should we at least reset the cursor?

Reply:
We have a trigger, but currently it is used for testing purposes. I do not see any problem in opening a cursor on the metadata pages until they are occupying the cache a lot. I will check other simple alternatives to release the cursor.